test/cpp/end2end/xds_end2end_test.cc

```diff
@@ -951,6 +976,8 @@ class AdsServiceImpl : public AggregatedDiscoveryService::Service,
   Cluster default_cluster_;
   std::map<std::string /*resource type*/, ResponseState>
       resource_type_response_state_;
+  std::map<std::string /*resource type*/, std::string /*error message*/>
```

Comment:
Instead of adding a second map for this, I suggest combining it with the previous map:

```
// Note: This changes ResponseState from an enum into a struct, with the original enum inside the struct.
struct ResponseState {
  enum State { NOT_SENT, SENT, ACKED, NACKED };
  State state = NOT_SENT;
  std::string error_message;
};

std::map<std::string /* type_url */, ResponseState> resource_type_response_state_;
```

Then change all the accessors to return `const ResponseState&`, so that we only need one accessor for each resource type instead of two:

```
const ResponseState& lds_response_state() {
  grpc_core::MutexLock lock(&ads_mu_);
  return resource_type_response_state_[kLdsTypeUrl];
}
```

Then, in each individual test, you can do something like this:

```
const auto& response_state = balancers_[0]->ads_service()->lds_response_state();
EXPECT_EQ(response_state.state, AdsServiceImpl::ResponseState::NACKED);
EXPECT_EQ(response_state.error_message, "Listener doesn\'t have ApiListener.");
```